### ISSUE TITLE
computing topology is time-expensive (slow)

### ISSUE THREAD

**Issue by mattijn** [mattijn/topojson#73]
The whole concept of creating a topology is based on finding paths (line segments) that are shared by two or more geometries. 

To find out if there are segments in common between two geometries I use the `shapely.ops.shared_paths` [function ](https://shapely.readthedocs.io/en/latest/manual.html#shared-paths) which is a Python interface to the GEOS `SharedPathsOp` [function](https://geos.osgeo.org/doxygen/classgeos_1_1operation_1_1sharedpaths_1_1SharedPathsOp.html).
But in the case of fairly complex geometries this function is time-expensive (slow).

Let me show this using an example. I use the boroughs boundaries of New York City available within geopandas.

```python
import geopandas as gpd
from shapely.ops import shared_paths

gdf = gpd.read_file(gpd.datasets.get_path('nybb'))
ax = gdf[gdf.BoroName!='Queens'].plot(color='lightblue')
gdf[gdf.BoroName=='Queens'].plot(ax=ax)
ax.set_axis_off()
```
![image](https://user-images.githubusercontent.com/5186265/80503456-2b107f80-8972-11ea-919f-3c230b4ea8c4.png)

For the purpose of this example I extract the exterior of the largest polygon from the Queens borough that encompass more than 16.000 coordinates. 

```python
geom = gdf.iloc[1].geometry[17].exterior
print(f'no. of coords in linestring: {len(geom.coords)}')
geom
```
![image](https://user-images.githubusercontent.com/5186265/80503511-41b6d680-8972-11ea-9b09-ac4a515ca0ec.png)

When comparing two complex linestrings that have segments in common using the `shapely.ops.shared_paths` function we will quickly realize this is very time-expensive: 

```python
%%timeit
shared_paths(geom, geom)
```

> `22.7 s ± 351 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`

Yes, you read it well. That is 22.7 seconds. For a single comparison. 

A lot of effort in this repo has been aimed at reducing the number of linestrings that should be compared against each other, but in the end, it is also the core of the repo: _'detection of shared paths between geometries'_. Sure, upon knowing the shared paths, a lot of of other things needs to be done to cast it into proper topojson format, but in time comparisons its peanuts.

Moreover, this GEOS function does not perform any better with broadcasting (shapely vs pygeos example [here](https://github.com/pygeos/pygeos/issues/73#issuecomment-552175579))

So where we stand: the GEOS `SharedPathsOp` function accessible by both shapely and pygeos is time-expensive (slow). 

Two open questions to conclude:
- Maybe a pure numpy approach can improve timings, especially considering broadcasting (did it before for the slow `shapely.ops.split` function now available as `topojson.ops.fast_split` [here](https://github.com/mattijn/topojson/blob/master/topojson/ops.py#L116:L168))?
- Maybe the concept of having a shared path function alike to create a topology is wrong?

**Comment by martinfleis** [comment]
NumPy approach would be likely significantly faster, but! it needs topologically correct input. By that I mean that you need shared vertices and edges to do that. Shared paths do not need any of it. Consider following example:

```py
from shapely.geometry import LineString
from shapely.ops import shared_paths
import geopandas as gpd

box1 = LineString([(0,0), (0,1), (1, 1), (1, 0), (0,0)])
box2 = LineString([(-1,0), (2,0), (2, -2), (-1, -2), (-1,0)])
gpd.GeoSeries([box1, box2]).plot(color=['r', 'b'], alpha=.5)

shared_paths(box1, box2).wkt

'GEOMETRYCOLLECTION (MULTILINESTRING EMPTY, MULTILINESTRING ((1 0, 0 0)))'
```
![image](https://user-images.githubusercontent.com/36797143/80588038-7f1d7180-8a0f-11ea-81b1-3947e9f76f50.png)

These two line strings are topological neighbours, so they should be recognised by TopoJSON as such, but they do not share even a single vertex.

Exactly the same issue is in `libpysal` library during the creation of spatial weights matrix. They offer both options - [Queen](https://pysal.org/libpysal/generated/libpysal.weights.Queen.html#libpysal.weights.Queen) or [Rook](https://pysal.org/libpysal/generated/libpysal.weights.Rook.html#libpysal.weights.Rook) weights based purely on shared vertices/edges, which is the recommended performant option. Then they also have [Fuzzy contiguity function](https://pysal.org/libpysal/generated/libpysal.weights.fuzzy_contiguity.html) based on shapely's `intersects`.

It sounds like a good model for TopoJSON as well. Defaulting to fast numpy algorithm, but optionally offer shared_paths.

**Comment by mattijn** [comment]
Yes, I'm aware of that. Forgot to include this info in the issue. I've considered a numpy route before and eventually sticked to the shapely shared_paths strategy because of the situation you mentioned. 

I like your suggestion to include both a numpy and shapely route. Still not completely sure about performance of a numpy-based function though. Opened a PR for testing: https://github.com/mattijn/topojson/pull/75.

I also don't really need the shared paths. I need the two coordinates that represent both of the segment that is shared. Given that info, maybe someone can think of yet another strategy

**Comment by martinfleis** [comment]
Hi, I did some tests with #76 and it seems that `Join` is no longer the bottleneck. I have used fairly complex (but still small) GeoDataFrame with ~4000 dense polygons with following timings:

```
topojson.Topology(gdf, shared_paths='dict')

extract 4.612934827804565
join 6.585726261138916
cut 16.98685908317566
dedup 14.114187955856323
hash 35.22837710380554
total 77.5300726890564
```

```
topo = topojson.Topology(gdf, shared_paths='shapely')

extract 4.71062707901001
join 97.8879702091217
cut 24.696655988693237
dedup 14.725230693817139
hash 47.4505250453949
total 189.47293210029602
```

I had to stop `numpy` version as it took ages. If you want to try the data, they're [here](https://www.dropbox.com/s/7m1skmeafz1jzs2/sample.geojson?dl=0) for some time.

**Comment by mattijn** [comment]
I've released version `1.0rc10`, which results in the following speedups for me:


| release | naturalearth_lowres | nybb | sample.geojson |
|---------------|---------------------------|---------------------------|---------------------------|
| **1.0rc10** (15 June 2020) | 8X (214 ms) | 20X (514 ms) | 20X (12.9 s) |

I'll close this issue for now.
